src/main/java/org/mockito/internal/util/KotlinInlineClassUtil.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2021 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+package org.mockito.internal.util;
+
+import org.mockito.internal.stubbing.answers.InvocationInfo;
+import org.mockito.invocation.InvocationOnMock;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class KotlinInlineClassUtil {
+    private static Class<Annotation> jvmInlineAnnotation;
+
+    static {
+        try {
+            jvmInlineAnnotation = (Class<Annotation>) Class.forName("kotlin.jvm.JvmInline");
+        } catch (ClassNotFoundException e) {
+            // Do nothing: kotlin is pre 1.5.0
+        }
+    }
+
+    // When mocking function, returning inline class, its return type is
+    // underlying type.
+    // So, `thenReturn` calls fails, because of non-compatible types.
+    public static boolean isInlineClassWithAssignableUnderlyingType(
+            Class<?> inlineClass, Class<?> underlyingType) {
+        try {
+            // Since 1.5.0 inline classes have @JvmInline annotation
+            if (jvmInlineAnnotation == null
+                    || !inlineClass.isAnnotationPresent(jvmInlineAnnotation)) return false;
+
+            // All inline classes have 'box-impl' method, which accepts
+            // underlying type and returns inline class.
+            inlineClass.getDeclaredMethod("box-impl", underlyingType);
```

Comment:
If all inline classes have the annotation, is this still necessary? If I am not mistaken, we can update this method to the following:

```java
public static boolean isInlineClassWithAssignableUnderlyingType(
            Class<?> inlineClass, Class<?> underlyingType) {
    return jvmInlineAnnotation != null
                     && inlineClass.isAnnotationPresent(jvmInlineAnnotation);
}
```

Reply:
Yes, otherwise, it would be possible to mock a function with different inline class:
```
inline class ICAny(val a: Any)
inline class ICInt(val i: Int)

fun foo(): ICInt

`when`(foo()).thenReturn(ICAny)
```
will try to unbox and then cast the result, failing.

Reply:
Oooh I see. Yes that makes sense now. Could we then update the comment to make that a bit clearer?

```java
// Make sure that the current inline class is also the class that is compatible with the underlying type.
// If we don't make this check, then we would potentially pass a mock of inline type A into a method
// that accepts inline type B.
```